metals/src/main/scala/scala/meta/internal/metals/LineListener.scala

```diff
@@ -0,0 +1,117 @@
+// This file contains adapted source code from tut, see NOTICE.md for LICENSE.
+// Original source: https://github.com/tpolecat/tut/blob/e692c74afe7cb9f144f464b97f100c11367c7dfa/modules/core/src/main/scala/tut/AnsiFilterStream.scala
+package scala.meta.internal.metals
+
+import java.lang.StringBuilder
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+import java.nio.CharBuffer
+import scala.collection.mutable
+
+/**
+ * Converts a stream of strings (with potential ANSI codes and newlines) into a callback for indvidual lines
+ * where each individual line has no newline \n characters.
+ *
+ * @param onLine The callback handler when a single line has been consumed from the input.
+ */
+class LineListener(onLine: String => Unit) {
+  private var buffer = new StringBuilder()
+
+  /** Clear buffered output. */
+  def flush(): Unit = {
+    if (buffer.length() > 0) {
+      onLine(buffer.toString())
+      buffer = new StringBuilder()
+    }
+  }
+
+  def appendString(text: String): this.type = {
+    appendBytes(
+      StandardCharsets.UTF_8.encode(CharBuffer.wrap(text.toCharArray()))
+    )
+  }
+
+  def appendBytes(bytes: ByteBuffer): this.type = {
+    val chars = StandardCharsets.UTF_8.decode(bytes)
+    while (chars.hasRemaining()) {
+      val ch = chars.get()
+      // Strategy is, accumulate values as long as we're in a non-terminal state, then either discard
+      // them if we reach T (which means we accumulated an ANSI escape sequence) or print them out if
+      // we reach F.
+      state.apply(ch) match {
+        case F =>
+          stack.foreach { i =>
+            onCharacter(i)
+          }
+          onCharacter(ch)
+          resetState()
+        case T =>
+          resetState()
+        case L =>
+          flush()
+          resetState()
+        case s =>
+          stack += ch
+          state = s
+      }
+    }
+    this
+  }
+
+  private def onCharacter(char: Char): Unit = {
+    if (char == '\n') {
+      flush()
+    } else {
+      buffer.append(char)
+    }
+  }
+
+  case class State(apply: Int => State)
```

Comment:
I've refactored the state machine into a separate file and given the states more readable names. I personally find the new approach easier to maintain and evolve since we're conceptually doing a simpler task, we filter out a few bytes from the process output and emit lines as soon as we find a `\n`. The previous implementation had multiple layers of conversions that were difficult to follow and extend without adding yet another full iteration on the character output stream.